test/system/src/test/java/io/pravega/test/system/ReaderCheckpointTest.java

```diff
@@ -0,0 +1,242 @@
+        readEventsAndVerify(startInclusive, endExclusive);
+
+        //initiate checkpoint100
+        Checkpoint checkPoint100 = readerGroup.initiateCheckpoint("batch100", executor).get();
+
+        //write and read events 100 to 200
+        startInclusive = 100;
+        endExclusive = 200;
+        log.info("Write events with range [{},{})", startInclusive, endExclusive);
+        writeEvents(IntStream.range(startInclusive, endExclusive).boxed().collect(Collectors.toList()));
+        readEventsAndVerify(startInclusive, endExclusive);
+
+        //reset to check point 100
+        readerGroup.resetReadersToCheckpoint(checkPoint100);
+        readEventsAndVerify(100, endExclusive);
+
+        //initiate checkpoint200
+        Checkpoint checkPoint200 = readerGroup.initiateCheckpoint("batch200", executor).get();
+
+        //write and read events 200 to 300
+        startInclusive = 200;
+        endExclusive = 300;
+        log.info("Write events with range [{},{})", startInclusive, endExclusive);
+        writeEvents(IntStream.range(startInclusive, endExclusive).boxed().collect(Collectors.toList()));
+        readEventsAndVerify(startInclusive, endExclusive);
+
+        //reset back to checkpoint 200
+        readerGroup.resetReadersToCheckpoint(checkPoint200);
+        readEventsAndVerify(200, endExclusive);
+
+        //reset back to checkpoint 100
+        readerGroup.resetReadersToCheckpoint(checkPoint100);
+        readEventsAndVerify(100, endExclusive);
+
+        readerGroupManager.deleteReaderGroup(READER_GROUP_NAME); //clean up
+    }
+
+    private void readEventsAndVerify(int startInclusive, int endExclusive) {
+        log.info("Read and Verify events between [{},{})", startInclusive, endExclusive);
+        List<EventRead<Integer>> eventsRead = readEvents();
+        verifyEvents(eventsRead, startInclusive, endExclusive);
+    }
+
+    private void verifyEvents(List<EventRead<Integer>> events, int startInclusive, int endExclusive) {
+
+        Supplier<Stream<Integer>> streamSupplier = () -> events.stream().map(i -> i.getEvent()).sorted();
+        IntSummaryStatistics stats = streamSupplier.get().collect(Collectors.summarizingInt(value -> value));
+
+        assertTrue("Check for first event", stats.getMin() == startInclusive);
+        assertTrue("Check for last event", stats.getMax() == endExclusive - 1);
+        //Check for missing events
+        assertEquals("Check for number of events", endExclusive - startInclusive, stats.getCount());
+        assertEquals("Check for duplicate events", endExclusive - startInclusive,
+                streamSupplier.get().distinct().count());
+    }
+
+    private <T extends Serializable> List<EventRead<T>> readEvents() {
+        List<EventRead<T>> events = new ArrayList<>();
+
+        try (ClientFactory clientFactory = ClientFactory.withScope(SCOPE, controllerURI);
+             EventStreamReader<T> reader = clientFactory.createReader("reader",
```

Comment:
I suggest we use multiple readers.

Reply:
updated the code to use multiple readers.